runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,219 @@
+package org.corfudb.runtime.collections;
+
+import com.google.protobuf.Message;
+
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nonnull;
+
+import org.corfudb.protocols.logprotocol.MultiObjectSMREntry;
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.SubscriberAlreadyExistsException;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.runtime.view.ObjectsView;
+import org.corfudb.runtime.view.StreamOptions;
+import org.corfudb.runtime.view.TableRegistry;
+import org.corfudb.runtime.view.stream.IStreamView;
+
+/**
+ *
+ * Created by hisundar on 04/28/2020.
+ */
+@Slf4j
+public class StreamManager {
+
+    private final int MAX_SUBSCRIBERS = 12; // Arbitrary sanity check value.
+
+    /**
+     * Corfu Runtime.
+     */
+    @Getter
+    private final CorfuRuntime runtime;
+
+    /**
+     * Map of StreamingSubscriptionThreads.
+     * <p>
+     * This is the map of all streaming subscriptions.
+     */
+    @Getter
+    private final Map<StreamListener, SubscriptionThread> subscriptions = new HashMap<>();
+
+    public StreamManager(@Nonnull CorfuRuntime runtime) {
+        this.runtime = runtime;
+    }
+
+    /**
+     * Subscribe to updates.
+     *
+     * @param streamListener   Client listener.
+     * @param namespace        Namespace of interest.
+     * @param tablesOfInterest Only updates from these tables will be returned.
+     * @param startAddress     Address to start the notifications from.
+     */
+    public synchronized <K extends Message, V extends Message, M extends Message>
+    void subscribe(@Nonnull StreamListener streamListener, @Nonnull String namespace,
```

Comment:
Since you implement pub-sub, it would be great if you could use https://www.reactive-streams.org/ instead of self-written protocols.
reactive-streams is a part of Java 9, so when we migrate to newer versions of java it can be used natively

Reply:
sure - we would be asked to move to Java 11 shortly anyway, but as of Java 8 this functionality is not available. Also this gives us fine grained control over how the streams should behave.

Reply:
you can add a dependency when we migrate to the new version of java we just get rid of that  dependency.
Also you write a lot of code which makes migration to reactive streams more and more complicated over time, we would end up with our custom version, now is the best time to migrate, it's cheap and save and takes couple of hours of work

Reply:
Well this implementation is 600 lines of lesser code compared to previous (existing). It's just one file, so can't think of an easier way to migrate to RxJava if we should decide to use it in the future. 

Also its not just dependency, RxJava internally uses a thread pool based on a publisher subscriber model. The use of separate publisher + subscriber is the root cause of the very bug which prompted this change, so we have to be careful about adding back a publisher-subscriber based model. For now, it makes sense to have something that works and more importantly does not break anything.

Reply:
I'm not talking about RxJva, only reactive-streams, it's much smaller

Reply:
reactive-streams is also based on a publisher-subscriber model with functionality for backpressure both of which are anti-patterns given the current use case where we have JVMs with low heap space needing to selectively poll. In the current form where we still have a big global TRANSACTION_STREAM anything other than simple threads processing one mutation at a time will hit issues. 

Now that is not to say reactive-streams should not be used. We need the stream tagging feature first. That will allow us to be selective in streaming at which point we would be able to have better control over the subscriptions and we can avoid OOMs. Until then we do need fine grained control over how stream listeners work. Using direct threads achieves just that in probably the simplest way possible.